### ISSUE TITLE
Handling dbus message data when implementing my own event loop

### ISSUE THREAD

**Issue by Noomaok** [Kistler-Group/sdbus-cpp#141]
Hi, I am posting this issue because I kind of run in a problem.

I am trying to implement my own event loop with boost asio to handle dbus events because my whole program use this to do asynchronous tasks.

I first used the example you provided in `using-sdbus-c++` with the xml file and the source code generator. Then I altered the code so that I don't run `connection->enterEventLoop()` but instead use the `getEventLoopPollData` function. I can receive some data from the client, but then I am a bit lost on own we are supposed to process it.
If I have to parse the data myself, then it's not very interesting to use this library in this way.
Is there a way to let the library parse the data when it's received ?

The code I used for the server is the following :
```cpp
#include <boost/asio.hpp>
#include <boost/asio/spawn.hpp>
#include <boost/bind/bind.hpp>

#include <iostream>

#include "Concatenator.h"

void watch_descriptor(boost::asio::yield_context yield, boost::asio::io_service& ios, int file_descriptor)
{
    boost::system::error_code error;

    auto desc = new boost::asio::posix::stream_descriptor(ios);
    desc->assign(file_descriptor);

    while (true)
    {
        uint8_t data[4096];
        const auto bytesRead = desc->async_read_some(boost::asio::buffer(data), yield[error]);

        if(error) {
            throw std::runtime_error("error: " + error.message());
        }

        std::cout << bytesRead << " bytes" << std::endl;
        
        // Process dbus data here
    }
    
}

int main()
{
    boost::asio::io_context asioContext;

    const char* serviceName = "org.sdbuscpp.concatenator";
    auto connection = sdbus::createSystemBusConnection(serviceName);

    const char* objectPath = "/org/sdbuscpp/concatenator";
    Concatenator concatenator(*connection, objectPath);

    auto data = connection->getEventLoopPollData();

    boost::asio::signal_set signals(asioContext, SIGINT, SIGTERM);
    signals.async_wait([&](auto, auto) {
        std::cout << "stop" << std::endl;
        asioContext.stop();
    });

    boost::asio::spawn(asioContext,
                       [&](boost::asio::yield_context yield) 
                       { watch_descriptor(yield, asioContext, data.fd); });

    //connection->enterEventLoop();
    asioContext.run();
}

```

**Comment by dleeds-cpi** [comment]
Integrating an external event loop is possible, but made somewhat more complicated by some questionable interface decisions made by the underlying `systemd/sd-bus` library. 
1. Poll `getEventLoopPollData()` at the end of _every_ event loop iteration (even for events not prompted by `sd-bus`). 
2. At each poll, ensure a FD watcher is watching for READ and/or WRITE events on `data.fd`.
3. If `data.timeout_usecs` does not equal `-1`, set a timer to expire at `timeout_usecs` on the monotonic clock (it is a time point, not a duration).
4. On any event from an FD watcher or timer, call `processPendingRequest()`.

**Comment by dleeds-cpi** [comment]
Here is a snippet from my solution:

```c++
dbus::dbus(unique_ptr<IConnection> conn)
{
    // Setup callbacks
    post_.set_callback([this](auto &post) { configure(); });
    io_.set_callback([this](auto &io, auto mask) { process(); });
    timer_.set_callback([this](auto &timer) { process(); });

    // Take ownership of the connection pointer
    attach(move(conn));
}

// Register for bus events to process
void dbus::configure()
{
    IConnection::PollData poll = conn_->getEventLoopPollData();

    if (poll.fd < 0) {
        io_.cancel();
    } else if (poll.fd != io_.get_fd() || event::io::mask_from_poll_events(poll.events) != io_.get_events()) {
        io_.listen(poll.fd, event::io::mask_from_poll_events(poll.events));
    }

    if (poll.timeout_usec == (uint64_t)-1) {
        timer_.cancel();
    } else {
        event::timer::expiration_type timeout(chrono::microseconds(poll.timeout_usec));
        timer_.set(timeout);
    }
}

// Process pending operations
void dbus::process()
{
    for (size_t i = 1; conn_->processPendingRequest() && i < DBUS_PROCESS_MAX; ++i);
}
```

**Comment by fwilmsh** [comment]
@dleeds-cpi I'm pretty much looking for the exact same thing as @Noomaok does, so thanks for providing your snippet. I think I do get the idea you are proposing, I can't put the source snippet into a greater perspective. Could you elaborate a little what types/API are used as `post_`, `io_` and `timer_`? Or is it just pseudo-code? :)

**Comment by dleeds-cpi** [comment]
@fwilmsh I made a C++ wrapper around the systemd `sd-event` library, and my `post`, `io`, and `timer` event loop sources reflect the behavior of the underlying `sd_event_post`, `sd_event_io`, and `sd_event_timer` event sources. See https://github.com/systemd/systemd/blob/main/src/systemd/sd-event.h#L89 Unfortunately, not all `sd-event` functionality is documented. The `post` event invokes a callback at the end of each event loop iteration.

**Comment by dleeds-cpi** [comment]
@fwilmsh I am not very familiar with `boost::asio`, but based on code inspection, you will need to add a timer, and also to update your timer and I/O listener on each event loop iteration (not just in the constructor).

**Comment by anthonybrandon** [comment]
I want to do something similar (using boost asio as eventloop) but I wonder if this is actually possible with the current API.
For one, using a custom event loop enterEventLoop will never be called, and therefore loopThreadId_ will never be set.
This needs to be set for tryCallMethodSynchronously to work.
Secondly `notifyEventLoopNewTimeout` will still be called, so I think a way is needed at least to wait on that fd and clear it.
Or am I overthinking things?

**Comment by sangelovic** [comment]
Hi there. Just to give heads-up... There is a release branch for the upcoming v2.0 release. #309 merged into that branch brought several improvements, internal simplifications and fixes which also relate to the external event loops. There is a new section in the tutorial which tries to explain in more detail how to hook sdbus-c++ up with external event loops: [also ](https://github.com/Kistler-Group/sdbus-cpp/blob/release/v2.0/docs/using-sdbus-c%2B%2B.md#using-sdbus-c-in-external-event-loops). A follow-up PR with complete integration of sd-event is expected to be submitted soon.
